### PR TITLE
add timestamps to Dolly::Document

### PR DESCRIPTION
**Pull request by seancookr**
This is currently a work in progress.

### CODE REVIEW

File: lib/dolly/document.rb [hunk 1]
```diff
@@ -19,6 +19,7 @@ def update_properties options = {}
       options.each do |property, value|
         send(:"#{property}=", value)
       end
+      updated_at = DateTime.now if has_updated_at?
```
Comment (seancookr) [inline lib/dolly/document.rb]: I dont want to hardcode this as DateTime and would rather take the class of the property.  Going to do that in the next commit

File: test/document_test.rb [hunk 2]
```diff
@@ -290,6 +291,12 @@ def setup
     end
   end
 
+  test 'set updated at' do
+    foo = FooBar.new 'id' => 'a', foo: 'ab'
+    foo.update_properties! foo: 'c'
+    assert_equal DateTime.now.to_s, foo.updated_at.to_s
```
Comment (seancookr) [inline test/document_test.rb]: I originally ommitted the `#to_s`, however when I did I recieved this

> FAIL (0:00:00.140) test_set_updated_at
          No visible difference in the DateTime#inspect output.
          You should look at the implementation of #== on DateTime or its members.

**Comment by seancookr** [comment]
@rubenrails, @javierg and I were discussing perhaps having all documents initially have `created_at` & `updated_at` properties, with a way to designate in the child if you would like to omit them.  What are your thoughts on this inclusion?

**Comment by rubenrails** [comment]
I totally agree. I had to implement that myself and I don't like having that repetition on each project. There seems to be some work done on that field but it isn't working. https://github.com/amco/dolly/issues/33

**Comment by javierg** [comment]
> There seems to be some work done on that field but it isn't working. #33

that is just handling timestamps as data type, remember in couch all data are JSON, so there is no actual data stamps, just strings and numbers

**Comment by rubenrails** [comment]
Ohh, that makes sense :+1:

**Comment by rubenrails** [comment]
Good job! Ship it

File: lib/dolly/document.rb [hunk 3]
```diff
@@ -1,9 +1,13 @@
 require "dolly/query"
 require "dolly/property"
+require 'dolly/timestamps'
+require 'active_model/callbacks'
 
 module Dolly
   class Document
     extend Dolly::Connection
+    extend ActiveModel::Callbacks
```
Comment (javierg) [inline lib/dolly/document.rb]: Noooooo!!!!!!

Comment (javierg) [inline lib/dolly/document.rb]: @smathy what do you think about this?

Comment (seancookr) [inline lib/dolly/document.rb]: @neojin you too!!

Comment (neojin) [inline lib/dolly/document.rb]: Sup guys - 

Isn't this the same path we went down with `couch_rest_adapter` ? ActiveModel will suck you in, which is fine, but I guess it really depends on your desired purpose.

ex) https://github.com/amco/couch_rest_adapter/blob/f366e14c3864c346e6d01debb435a64966cae805/lib/couch_rest_adapter.rb#L17-L20

Comment (seancookr) [inline lib/dolly/document.rb]: @neojin We are noticing that in our application logic (in more than 1 app, all of them that use dolly in fact) that we have a legitimate use for timestamps on documents.  My proposal is to abstract this to the gem layer so it doesn't have to be included in every model that inherits from `Dolly::Document`.  This is just on implementation, and Im all ears on if there are better implementations to have timestamps without including `ActiveModel::Callbacks`

Comment (javierg) [inline lib/dolly/document.rb]: My proposal is to have a class method like `timestamps!` that you can call on the class definition, to add them

Comment (neojin) [inline lib/dolly/document.rb]: I like @javierg 's idea. It is less magic, which is a good thing. You could even go as far as telling it which fields are timestamps, with some sensible defaults.

`timestamps! created: :foo, updated: :bar # default is :created_at, :updated_at`

Comment (seancookr) [inline lib/dolly/document.rb]: @neojin noted, and kindly considered. Ill look into it

**Comment by netlogics** [comment]
Code looks good. Is it decided that we are going to use active model? If not we could roll our own. Here are some references
https://github.com/apotonick/hooks
https://github.com/auser/backcall

File: lib/dolly/timestamps.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+require 'active_support/concern'
+require 'active_model/callbacks'
```
Comment (javierg) [inline lib/dolly/timestamps.rb]: I still think we should not add active model callbacks

**Comment by seancookr** [comment]
@netlogics thanks for those resources, Ill look into them

**Comment by netlogics** [comment]
@seancookr np. Go code ninja and roll your own so we can put this debate to rest.

**Comment by javierg** [comment]
http://www.samuelmullen.com/2013/05/the-problem-with-rails-callbacks/

File: lib/dolly/timestamps.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+require 'active_support/concern'
```
Comment (javierg) [inline lib/dolly/timestamps.rb]: you can remove the concerns require, as you are not using any of its methods.

Comment (javierg) [inline lib/dolly/timestamps.rb]: or... you can wrap the `tiemstamps!` method on a `classMethods` block and keep the Concern module

File: lib/dolly/document.rb [hunk 6]
```diff
@@ -50,6 +52,8 @@ def save
       obj = JSON::parse response.parsed_response
       doc['_rev'] = obj['rev'] if obj['rev']
       obj['ok']
+      set_created_at if respond_to? :set_created_at
+      set_updated_at if respond_to? :set_updated_at
```
Comment (javierg) [inline lib/dolly/document.rb]: I think this 2 lines should go before the response call

File: lib/dolly/timestamps.rb (hunk 5 above)
Comment (seancookr) [inline lib/dolly/timestamps.rb]: no need as concerns are already in the dolly document class

File: lib/dolly/document.rb (hunk 6 above)
Comment (seancookr) [inline lib/dolly/document.rb]: ok

File: lib/dolly/timestamps.rb (hunk 5 above)
Comment (javierg) [inline lib/dolly/timestamps.rb]: ok, so oyu can use the `classMethods` block and use included in the Document class

**Comment by javierg** [comment]
:shipit:

**Pull request merged**